doc/configuration.md

```diff
@@ -0,0 +1,288 @@
+- **Config Validation**: Custom enum types for for fields such as `pub enum HostnameProvisioner` validate that user-set config settings match the supported options. Invalid values will cause deserialization to fail.
+- **Built-in Defaults**: The system defines defaults directly in the code using Rust's `Default` trait, eliminating the need for a separate default config file.
+- **Merging of Configurations**: The `load()` and `merge()` methods merge multiple sources of configuration data. Defaults are loaded first, then overridden by values from a config file, and finally by CLI-specified configurations.
+
+## Config File Structure
+**Format**: TOML
+
+- The configuration relies on default values defined in the source code (`config.rs`).
+- Users can override these defaults by providing a TOML configuration file.
+- A custom configuration file can be passed via the CLI or added by the user.
+
+### CLI Parameters
+Example: `--config /etc/azure-init/`
+
+## Configuration Hierarchy for `azure-init`
+
+This document outlines the configuration system for `azure-init`, which allows flexibility by merging settings from multiple sources. Configuration can be set via a single file, a directory containing multiple files, or default values defined in the code.
+
+### Configuration Loading Order
+
+#### 1. CLI Override (`--config`)
+   - The `--config` flag specifies the path to load the configuration from, which can be either a single file or a directory.
+     - **File:** If the path points to a file, only that file is loaded.
+     - **Directory:** If the path points to a directory, a series of `.toml` files in the specified directory will be loaded and merged based on specific rules.
+   - **Example:** `azure-init --config /path/to/custom-config.toml`
+
+#### 2. Directory Loading Logic
+   - If the `--config` parameter points to a directory, `azure-init` follows this hierarchy:
+     - First, it looks for a base file named `azure-init.toml` in the directory.
+     - Then, it merges any additional `.toml` files in a `.d` subdirectory within the specified directory, in lexicographical order.
+     ```text
+     /etc/azure-init/
+     ├── azure-init.toml            # Base configuration
+     └── .d/
+         ├── 01-network.toml        # Additional network configuration
+         ├── 02-ssh.toml            # Additional SSH configuration
+         └── 99-overrides.toml      # Final overrides
+     ```
+
+#### 3. Defaults in Code
+   - If neither a file nor a directory is provided, `azure-init` falls back to using the default values specified in `config.rs`.
+
+### Example 1: Single Configuration File
+**Command:**
+```sh
+azure-init --config /path/to/custom-config.toml
+```
+**Order of Merging:**
+1. Loads `/path/to/custom-config.toml`.
+2. Applies any CLI overrides if present.
+3. Fills in missing values with defaults from `config.rs`.
```

Comment:
based on the description, I expect the order of precedence (lowest to highest) to be:

(1) compile-time defaults
(2a) custom configuration (e.g. azure-init.toml)
(2b) custom configurations found in azure-init.toml.d, in lexicographical order of file names
(3) cli options

I think maybe it would help to clarify it as such.  "fills in missing values" feels a bit mismatched in terms of "merging". If it were an order of operations for merging, the defaults should like at the top and future merges overwrite it.